featurebyte/service/tile_manager.py

```diff
@@ -60,39 +61,80 @@ async def generate_tiles_on_demand(
             list of TileSpec, temp_entity_table to update the feature store
         progress_callback: Optional[Callable[[int, str | None], Coroutine[Any, Any, None]]]
             Optional progress callback function
+
+        Raises
+        ------
+        Exception
+            If any task failed
         """
         num_jobs = len(tile_inputs)
+        processed = 0
+
+        async def _progress_callback() -> None:
+            nonlocal processed
+            processed += 1
+            if progress_callback:
+                pct = int(100 * processed / num_jobs)
+                await progress_callback(pct, f"Computed {processed} out of {num_jobs} tiles")
+
         if progress_callback:
-            await progress_callback(0, f"0/{num_jobs} completed")
-
-        for index, (tile_spec, entity_table) in enumerate(tile_inputs):
-            tic = time.time()
-            await self.generate_tiles(
-                session=session,
-                tile_spec=tile_spec,
-                tile_type=TileType.OFFLINE,
-                start_ts_str=None,
-                end_ts_str=None,
+            await progress_callback(0, "Computing tiles on demand")
+        tasks = []
+        for tile_spec, entity_table in tile_inputs:
+            tasks.append(
+                asyncio.create_task(
+                    self._generate_tiles_on_demand_for_tile_spec(
+                        session=session,
+                        tile_spec=tile_spec,
+                        entity_table=entity_table,
+                        progress_callback=_progress_callback,
+                    )
+                )
             )
-            logger.debug(
-                "Done generating tiles",
-                extra={"tile_id": tile_spec.tile_id, "duration": time.time() - tic},
+        try:
+            await asyncio.gather(*tasks)
+        except Exception:
+            logger.error(
+                "Canceling all other tasks because at least one task failed",
+                exc_info=True,
             )
+            for task in tasks:
+                task.cancel()
+            raise
 
-            tic = time.time()
-            await self.update_tile_entity_tracker(
-                session=session, tile_spec=tile_spec, temp_entity_table=entity_table
-            )
-            logger.debug(
-                "Done update_tile_entity_tracker",
-                extra={"tile_id": tile_spec.tile_id, "duration": time.time() - tic},
-            )
+    async def _generate_tiles_on_demand_for_tile_spec(
+        self,
+        session: BaseSession,
+        tile_spec: TileSpec,
+        entity_table: str,
+        progress_callback: Optional[Callable[[], Coroutine[Any, Any, None]]] = None,
+    ) -> None:
+        tic = time.time()
+        if not session.is_threadsafe():
+            session = await session.clone()
```

Comment:
Spark hangs without using different session objects across tasks.

Reply:
If I understood it correctly, is this the root cause of the Spark hangs occasionally?

Reply:
It shouldn't be since we don't cache spark session when getting a session from SessionManager. Is spark hanging still occurring in tests?

Reply:
I did not encounter spark hanging issue for a while.